File: sessao-5/tutorial-11-multiarquivo-agentes/exemplos/gerado/carrinho.py
"""
Saída do agente de IA — Módulo de Carrinho de Compras
Referência: Tutorial 11 — Geração multi-arquivo com agentes
Execute: python3 carrinho.py

ATENÇÃO: saída de agente após mudança multi-arquivo — inconsistência cross-file presente.
O agente recebeu a tarefa "adicionar suporte a cupom de desconto ao carrinho".
Ele atualizou precificacao.py corretamente (nova assinatura com 'cupom'),
mas NÃO atualizou os chamadores neste arquivo.

INCONSISTÊNCIA CROSS-FILE: a linha abaixo (em finalizar_carrinho) usa a assinatura
ANTIGA de calcular_total:

    total = calcular_total(itens)          ← ANTIGA (sem cupom) — não atualizada

A assinatura NOVA em precificacao.py exige:

    total = calcular_total(itens, cupom)   ← NOVA (com cupom obrigatório)

Em produção, finalizar_carrinho levantaria TypeError: "missing 1 required positional
argument: 'cupom'". Cada arquivo roda sua própria demo sem erro — a inconsistência
só aparece ao revisar o diff de ambos juntos. Veja exemplos/diff-comentado.md.
"""

from dataclasses import dataclass, field
from typing import Optional


# ─── Entidades (replicadas para manter auto-contenção deste arquivo) ──────────
# Em um módulo real estas viriam de precificacao.py; aqui são redeclaradas para
# que carrinho.py rode de forma independente.

@dataclass
class ItemCarrinho:
    produto_id:     str
    descricao:      str
    preco_unitario: float
    quantidade:     int


@dataclass
class Cupom:
    codigo:              str
    percentual_desconto: float


@dataclass
class Carrinho:
    id:     str
    itens:  list[ItemCarrinho] = field(default_factory=list)
    cupom:  Optional[Cupom]   = None


# ─── Operações de carrinho ────────────────────────────────────────────────────

def adicionar_item(carrinho: Carrinho, item: ItemCarrinho) -> None:
    """Adiciona um item ao carrinho. Itens com mesmo produto_id têm quantidade somada."""
    for existente in carrinho.itens:
        if existente.produto_id == item.produto_id:
            existente.quantidade += item.quantidade
            return
    carrinho.itens.append(item)


def remover_item(carrinho: Carrinho, produto_id: str) -> None:
    """Remove o item com o produto_id fornecido. Silencioso se não existir."""
    carrinho.itens = [i for i in carrinho.itens if i.produto_id != produto_id]


def aplicar_cupom(carrinho: Carrinho, cupom: Cupom) -> None:
    """Associa um cupom ao carrinho, substituindo qualquer cupom anterior."""
    carrinho.cupom = cupom


def _subtotal_item(item: ItemCarrinho) -> float:
    return round(item.preco_unitario * item.quantidade, 2)


# ── Função auxiliar local (NÃO atualizada pelo agente) ───────────────────────
# Esta versão de _calcular_total_local foi copiada da versão ANTERIOR de
# calcular_total (precificacao.py antes da mudança) e não recebe cupom.
# Em produção o código chamaria calcular_total de precificacao.py assim:
#
#   from precificacao import calcular_total
#   total = calcular_total(itens)      ← CHAMADA ANTIGA — falta o argumento cupom
#
# A nova assinatura exige: calcular_total(itens, cupom)
# ⚠️  INCONSISTÊNCIA CROSS-FILE: este módulo não foi atualizado pelo agente.

def _calcular_total_local(itens: list[ItemCarrinho]) -> float:
    """Cálculo de total SEM suporte a cupom (versão anterior — não atualizada)."""
    if not itens:
        raise ValueError("A lista de itens não pode ser vazia")
    return round(sum(_subtotal_item(i) for i in itens), 2)


def finalizar_carrinho(carrinho: Carrinho) -> dict:
    """
    Finaliza o carrinho e retorna um resumo com total.

    BUG LATENTE: em produção isto chamaria calcular_total de precificacao.py.
    Como o agente atualizou a assinatura mas não atualizou este chamador,
    a chamada seria:

        total = calcular_total(carrinho.itens)   ← TypeError: missing 'cupom'

    O resultado: carrinho.cupom é ignorado — o desconto nunca é aplicado.
    """
    total = _calcular_total_local(carrinho.itens)   # ← usa versão antiga (sem cupom)

    return {
        "carrinho_id": carrinho.id,
        "qtd_itens":   sum(i.quantidade for i in carrinho.itens),
        "total":       total,
        "cupom":       carrinho.cupom.codigo if carrinho.cupom else None,
        # ⚠️  total não reflete o desconto do cupom — inconsistência cross-file
    }


def formatar_carrinho(carrinho: Carrinho) -> str:
    """Formata o estado atual do carrinho para exibição."""
    linhas = [f"Carrinho {carrinho.id}:"]
    for item in carrinho.itens:
        linhas.append(
            f"  [{item.produto_id}] {item.descricao}: "
            f"R$ {item.preco_unitario:.2f} x {item.quantidade} = "
            f"R$ {_subtotal_item(item):.2f}"
        )
    bruto = sum(_subtotal_item(i) for i in carrinho.itens)
    linhas.append(f"  Subtotal: R$ {bruto:.2f}")
    if carrinho.cupom:
        linhas.append(f"  Cupom: {carrinho.cupom.codigo} (não aplicado — inconsistência)")
    return "\n".join(linhas)


# ─── Execução de demonstração ─────────────────────────────────────────────────

if __name__ == "__main__":
    print("=== Carrinho de Compras (saída do agente — chamador não atualizado) ===\n")

    carrinho = Carrinho(id="C-2026-001")

    adicionar_item(carrinho, ItemCarrinho("P001", "Teclado mecânico",  349.90, 1))
    adicionar_item(carrinho, ItemCarrinho("P002", "Mouse sem fio",      89.90, 2))
    adicionar_item(carrinho, ItemCarrinho("P003", "Mousepad XL",         49.90, 1))

    print(formatar_carrinho(carrinho))
    print()

    cupom_bv = Cupom(codigo="BEMVINDO10", percentual_desconto=0.10)
    aplicar_cupom(carrinho, cupom_bv)
    print(f"Cupom aplicado: {carrinho.cupom.codigo} (-10%)")

    resumo = finalizar_carrinho(carrinho)
    print(f"\nResumo final:")
    print(f"  Carrinho: {resumo['carrinho_id']}")
    print(f"  Itens:    {resumo['qtd_itens']}")
    print(f"  Cupom:    {resumo['cupom']}")
    print(f"  Total:    R$ {resumo['total']:.2f}  ← ERRADO: desconto de 10% não foi aplicado")

    bruto = sum(_subtotal_item(i) for i in carrinho.itens)
    esperado = round(bruto * 0.90, 2)
    print(f"  Esperado: R$ {esperado:.2f}  (com 10% de desconto BEMVINDO10)")

    print()
    print("A inconsistência cross-file: calcular_total em precificacao.py exige 'cupom'")
    print("mas finalizar_carrinho chama a versão antiga sem esse argumento.")
    print("O cupom está no carrinho mas o desconto nunca chega ao total.")

File: sessao-5/tutorial-11-multiarquivo-agentes/exemplos/gerado/carrinho.ts
/**
 * Saída do agente de IA — Módulo de Carrinho de Compras
 * Referência: Tutorial 11 — Geração multi-arquivo com agentes
 * Execute: npx ts-node carrinho.ts
 *
 * ATENÇÃO: saída de agente após mudança multi-arquivo — inconsistência cross-file presente.
 * O agente recebeu a tarefa "adicionar suporte a cupom de desconto ao carrinho".
 * Ele atualizou precificacao.ts corretamente (nova assinatura com 'cupom'),
 * mas NÃO atualizou os chamadores neste arquivo.
 *
 * INCONSISTÊNCIA CROSS-FILE: finalizarCarrinho chama internamente calcularTotal(itens)
 * sem o argumento 'cupom'. Em um build integrado, o TypeScript acusaria:
 *   TS2554: Expected 2 arguments, but got 1.
 * O cupom está no Carrinho mas o desconto nunca chega ao total calculado.
 * Cada arquivo roda sua própria demo sem erro — a inconsistência só aparece
 * ao revisar o diff de ambos juntos. Veja exemplos/diff-comentado.md.
 */

// ─── Interfaces (replicadas para auto-contenção deste arquivo) ────────────────

interface ItemCarrinho {
  produtoId:     string;
  descricao:     string;
  precoUnitario: number;
  quantidade:    number;
}

interface Cupom {
  codigo:             string;
  percentualDesconto: number;
}

interface Carrinho {
  id:    string;
  itens: ItemCarrinho[];
  cupom: Cupom | null;
}

interface ResumoCarrinho {
  carrinhoId: string;
  qtdItens:   number;
  total:      number;
  cupom:      string | null;
}

// ─── Operações de carrinho ────────────────────────────────────────────────────

function criarCarrinho(id: string): Carrinho {
  return { id, itens: [], cupom: null };
}

function adicionarItem(carrinho: Carrinho, item: ItemCarrinho): void {
  const existente = carrinho.itens.find(i => i.produtoId === item.produtoId);
  if (existente) {
    existente.quantidade += item.quantidade;
  } else {
    carrinho.itens.push({ ...item });
  }
}

function removerItem(carrinho: Carrinho, produtoId: string): void {
  carrinho.itens = carrinho.itens.filter(i => i.produtoId !== produtoId);
}

function aplicarCupom(carrinho: Carrinho, cupom: Cupom): void {
  carrinho.cupom = cupom;
}

function _subtotalItem(item: ItemCarrinho): number {
  return Math.round(item.precoUnitario * item.quantidade * 100) / 100;
}

// ── Função auxiliar local (NÃO atualizada pelo agente) ─────────────────────
// Esta versão foi copiada da lógica ANTERIOR (antes da mudança do agente).
// Em produção o código chamaria calcularTotal de precificacao.ts assim:
//
//   import { calcularTotal } from "./precificacao";
//   const total = calcularTotal(carrinho.itens);   // ← CHAMADA ANTIGA — falta cupom
//
// A nova assinatura exige: calcularTotal(itens, cupom)
// ⚠️  INCONSISTÊNCIA CROSS-FILE: este módulo não foi atualizado pelo agente.

function _calcularTotalLocal(itens: ItemCarrinho[]): number {
  if (itens.length === 0) throw new Error("A lista de itens não pode ser vazia");
  const bruto = itens.reduce((acc, i) => acc + _subtotalItem(i), 0);
  return Math.round(bruto * 100) / 100;
}

function finalizarCarrinho(carrinho: Carrinho): ResumoCarrinho {
  /**
   * BUG LATENTE: em produção isto chamaria calcularTotal de precificacao.ts.
   * Como o agente atualizou a assinatura mas não atualizou este chamador,
   * a chamada seria:
   *
   *   const total = calcularTotal(carrinho.itens);  // TS2554: Expected 2 args, got 1
   *
   * O resultado: carrinho.cupom é ignorado — o desconto nunca é aplicado.
   */
  const total = _calcularTotalLocal(carrinho.itens);  // ← usa versão antiga (sem cupom)

  return {
    carrinhoId: carrinho.id,
    qtdItens:   carrinho.itens.reduce((acc, i) => acc + i.quantidade, 0),
    total,
    cupom:      carrinho.cupom?.codigo ?? null,
    // ⚠️  total não reflete o desconto do cupom — inconsistência cross-file
  };
}

function formatarCarrinho(carrinho: Carrinho): string {
  const linhas: string[] = [`Carrinho ${carrinho.id}:`];
  for (const item of carrinho.itens) {
    linhas.push(
      `  [${item.produtoId}] ${item.descricao}: ` +
      `R$ ${item.precoUnitario.toFixed(2)} x ${item.quantidade} = ` +
      `R$ ${_subtotalItem(item).toFixed(2)}`
    );
  }
  const bruto = carrinho.itens.reduce((acc, i) => acc + _subtotalItem(i), 0);
  linhas.push(`  Subtotal: R$ ${bruto.toFixed(2)}`);
  if (carrinho.cupom) {
    linhas.push(`  Cupom: ${carrinho.cupom.codigo} (não aplicado — inconsistência)`);
  }
  return linhas.join("\n");
}

// ─── Execução de demonstração ─────────────────────────────────────────────────

const carrinho = criarCarrinho("C-2026-001");

adicionarItem(carrinho, { produtoId: "P001", descricao: "Teclado mecânico",  precoUnitario: 349.90, quantidade: 1 });
adicionarItem(carrinho, { produtoId: "P002", descricao: "Mouse sem fio",      precoUnitario:  89.90, quantidade: 2 });
adicionarItem(carrinho, { produtoId: "P003", descricao: "Mousepad XL",        precoUnitario:  49.90, quantidade: 1 });

console.log("=== Carrinho de Compras (saída do agente — chamador não atualizado) ===\n");
console.log(formatarCarrinho(carrinho));
console.log();

const cupomBv: Cupom = { codigo: "BEMVINDO10", percentualDesconto: 0.10 };
aplicarCupom(carrinho, cupomBv);
console.log(`Cupom aplicado: ${carrinho.cupom!.codigo} (-10%)`);

const resumo = finalizarCarrinho(carrinho);
console.log("\nResumo final:");
console.log(`  Carrinho: ${resumo.carrinhoId}`);
console.log(`  Itens:    ${resumo.qtdItens}`);
console.log(`  Cupom:    ${resumo.cupom}`);

const bruto = carrinho.itens.reduce((acc, i) => acc + _subtotalItem(i), 0);
const esperado = Math.round(bruto * 0.90 * 100) / 100;
console.log(`  Total:    R$ ${resumo.total.toFixed(2)}  <- ERRADO: desconto de 10% não foi aplicado`);
console.log(`  Esperado: R$ ${esperado.toFixed(2)}  (com 10% de desconto BEMVINDO10)`);

console.log();
console.log("A inconsistência cross-file: calcularTotal em precificacao.ts exige 'cupom'");
console.log("mas finalizarCarrinho chama a versão antiga sem esse argumento.");
console.log("O cupom está no carrinho mas o desconto nunca chega ao total.");
